### PR TITLE
[NativeAOT] Support library mode with NativeAOT on iOS-like platforms

### LINKED ISSUE #88737: NativeAOT library mode on iOS-like platforms
As documented in https://github.com/royalapplications/beyondnet/issues/34#issuecomment-1632510996 there's a demand to get NativeAOT library mode working on iOS-like platforms. This is specifically for `net8.0` TFM and NOT the Xamarin `net8.0-ios` TFM which may run into similar issues but requires fixes in the Xamarin SDK. It currently requires a couple of workaround and we should look into fixing the underlying issues:

- Linking step doesn't use correct `-sysroot`/`-target`. We currently bundle the apps through AppleAppBuilder in dotnet/runtime, and a custom bundler in Xamarin. However, for dylib builds the NativeAOT linker step on its own is mostly sufficient. We just need to specify proper `-target` (eg. `apple-arm64-ios11.0` or similar) and sysroot. The sysroot can be obtained through `xcrun`, or we can simply execute the whole linker using `xcrun -sdk iphoneos clang <options>` command to achieve the same result.
- There's a conflict with the new `mobile-librarybuilder-experimental` workload which currently needs to be disabled with `_IsAppleMobileLibraryMode=false`. We should disable it automatically for NativeAOT publishing.

### PR DESCRIPTION
**Pull request by ivanpovazan**
This PR enables support for building libraries for iOS-like platforms with NativeAOT.

## Changes

- Extended NativeAOT build integration targets to support building libraries for iOS-like platforms
- Adapted `AppleAppBuilder` to support bundling in library mode for both MonoAOT and NativeAOT
- Adapted `AppleAppBuilder` to support bundling in library mode when targeting iOS-like devices
- Adapted `src/tests/FunctionalTests/iOS/Simulator/LibraryMode` for testing the library mode on `iossimulators-arm64/x64` with NativeAOT - currently this only works **locally** please see comments in the project file for more info
    - https://github.com/dotnet/runtime/issues/93447)
- Added `src/tests/FunctionalTests/iOS/Device/LibraryMode` for testing the library mode on `ios-arm64` for both MonoAOT and NativeAOT

## Additional notes

To test **locally** one can execute the following from the repo root directory:
- Test `MonoAOT` library mode on `ios-arm64`:
    ```
    ./dotnet.sh build /t:Test -c Debug /p:TargetOS=ios /p:TargetArchitecture=arm64 src/tests/FunctionalTests/iOS/Device/LibraryMode/iOS.Device.LibraryMode.Test.csproj /p:RuntimeFlavor=Mono
    ```
    
- Test `NativeAOT` library mode on `ios-arm64`:
    ```
    ./dotnet.sh build /t:Test -c Debug /p:TargetOS=ios /p:TargetArchitecture=arm64 src/tests/FunctionalTests/iOS/Device/LibraryMode/iOS.Device.LibraryMode.Test.csproj /p:UseNativeAOTRuntime=true /p:RunAOTCompilation=false /p:RuntimeFlavor=CoreCLR
    ```
    
## Documentation

- [x] https://github.com/dotnet/runtime/issues/96742

--- 
Fixes https://github.com/dotnet/runtime/issues/88737

### CODE REVIEW

File: src/tasks/AppleAppBuilder/Xcode.cs [hunk 1]
```diff
@@ -330,8 +340,30 @@ public void CreateXcodeProject(string projectName, string cmakeDirectoryPath)
 
         if (isLibraryMode)
         {
-            string dylibName = $"lib{projectName}.dylib";
-            cmakeLists = cmakeLists.Replace("%DYLIB_PATH%", $"{Path.Combine(binDir, dylibName)}");
+            string libraryPath;
+            // TODO: unify MonoAOT and NativeAOT library paths
+            // Current differences:
+            // - NativeAOT produces {ProjectName}.dylib, while MonoAOT produces lib{ProjectName}.dylib
+            // - NativeAOT places the library in the 'workspace' location ie 'publish' folder, while MonoAOT places it in 'binDir' ie 'AppBundle'
```
Comment (rolfbjarne) [inline src/tasks/AppleAppBuilder/Xcode.cs:347]: A .dylib isn't usable for the end developer, you should probably produce a .framework instead.

**Comment by ivanpovazan** [comment]
Failures are related:
1. Since I introduced `VerifyOSSpecificTooling` target, it seems I broke some dependency with tooling in the build
2. Building on Helix for the library mode tests seem to be broken

File: src/tasks/AppleAppBuilder/Xcode.cs (hunk 1 above)
Comment (ivanpovazan) [reply src/tasks/AppleAppBuilder/Xcode.cs:347]: Thank you, this meant to be an "initial" support for the library mode. We will definitely consider going with `.framework` approach.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets [hunk 2]
```diff
@@ -173,7 +209,9 @@ The .NET Foundation licenses this file to you under the MIT license.
       <LinkerArg Include="@(NativeFramework->'-framework %(Identity)')" Condition="'$(_IsApplePlatform)' == 'true'" />
       <LinkerArg Include="-Wl,--eh-frame-hdr" Condition="'$(_IsApplePlatform)' != 'true'" />
     </ItemGroup>
+  </Target>
 
+  <Target Name="VerifyOSSpecificTooling">
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:214]: Why do we need to split this? More places than just tests.singlefile.targets refer to the existing target (you can search github). I would rather not break them if I can help it.

Comment (ivanpovazan) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:214]: I made the split for two reasons:
1. To avoid moving the code that probes OS-specific tooling to the beginning of the `SetupOSSpecificProps`
    - This is required because we need `xcrun` to be verified before populating all the Apple-specific properties introduced in this PR
2. To separate the logic in `SetupOSSpecificProps` as it is: 
    - Setting up various properties and items, and 
    - Verifying the existence of OS-specific tools that some of these properties and items depend on

I could remove the newly added target `VerifyOSSpecificTooling` and reorganize `SetupOSSpecificProps` to make all this work together, but I think it is cleaner to have a separate target. What do you think?

Comment (MichalStrehovsky) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:214]: > To avoid moving the code that probes OS-specific tooling to the beginning

I might be missing something, but can we just take lines 284-294 and paste them before line 60? The only "problematic" property would be `_CommandProbe` because it's not defined there, but we can just replace `$(_CommandProbe)` with `command -v`. We already don't use `$(_CommandProbe)` with Apple specific stuff because cross compilation from non-Apple to Apple platform is an academic scenario.

> I think it is cleaner to have a separate target

`VerifyOSSpecificTooling` would still be doing stuff like compiling System.Globalization.Native for StaticICULinking and `SetupOSSpecificProps` would still be checking for a valid SDK with xcrun. It doesn't feel like it moves the needle. Many 3rd party repos depend on this target and this is not a good reason to break them.

Comment (ivanpovazan) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:214]: I see, thank you for the suggestion and clarification! 
I will remove the `VerifyOSSpecificTooling` and adjust the existing target accordingly.

File: global.json [hunk 3]
```diff
@@ -9,7 +9,7 @@
   },
   "msbuild-sdks": {
     "Microsoft.DotNet.Arcade.Sdk": "9.0.0-beta.23559.3",
-    "Microsoft.DotNet.Helix.Sdk": "9.0.0-beta.23559.3",
+    "Microsoft.DotNet.Helix.Sdk": "9.0.0-beta.23563.1",
```
Comment (ivanpovazan) [inline global.json:12]: This will be reverted, I bumped the version only to test the required fix for deep signing introduced in: https://github.com/dotnet/arcade/pull/14164

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets [hunk 4]
```diff
@@ -130,13 +175,15 @@ The .NET Foundation licenses this file to you under the MIT license.
       <LinkerArg Include="-gz=zlib" Condition="'$(CompressSymbols)' != 'false'" />
       <LinkerArg Include="-fuse-ld=$(LinkerFlavor)" Condition="'$(LinkerFlavor)' != ''" />
       <LinkerArg Include="@(NativeLibrary)" />
-      <LinkerArg Include="--sysroot=$(SysRoot)" Condition="'$(SysRoot)' != ''" />
-      <LinkerArg Include="--target=$(TargetTriple)" Condition="'$(_IsApplePlatform)' != 'true' and '$(TargetTriple)' != ''" />
+      <LinkerArg Include="--sysroot=$(SysRoot)" Condition="'$(SysRoot)' != '' and '$(_iOSLibraryMode)' != 'true'" />
+      <LinkerArg Include="-isysroot $(SysRoot)" Condition="'$(SysRoot)' != '' and '$(_iOSLibraryMode)' == 'true'" />
```
Comment (ivanpovazan) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:179]: @jkotas regarding our discussion in https://github.com/dotnet/runtime/pull/94999#discussion_r1420504285 is there anything you would like me to address in this PR?

**Comment by jkotas** [comment]
I am wondering why we need so much extra stuff for building iOS-like libraries. Is it because all this extra stuff comes from Xamarin iOS when building executables?

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets (hunk 4 above)
Comment (jkotas) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:179]: I do not think it makes sense for `SysRoot` to mean SDK root (isysroot) when building iOS-like libraries and a classic SysRoot for all other Apple targets.

I think overriding SDK root is an advanced scenario for all Apple targets. If we want `SysRoot` on Apple platforms to mean SDK root, this condition should be `_IsApplePlatform`.

(Alternatively, if we want to keep `SysRoot` as a classic SysRoot on Apple platforms, we should invent a property with a different name for the SDK root. Again, it should be applicable to all Apple platforms.)

**Comment by ivanpovazan** [comment]
> I am wondering why we need so much extra stuff for building iOS-like libraries. Is it because all this extra stuff comes from Xamarin iOS when building executables?

Yes, when building executables, apple SDK, target triple and all apple-specific build parameters, are resolved by Xamarin SDK and tooling that creates an application bundle. 
We have our internal `AppleAppBuilder` https://github.com/dotnet/runtime/tree/main/src/tasks/AppleAppBuilder having a minimal set of such capabilities, which we use for CI and local testing.

**Comment by jkotas** [comment]
Is `_iOSLibraryMode` the right name for this then? Based on your description, it is a mode without dependency on Xamarin SDK or Apple builder. We happen to use for building dynamic libraries in this PR, but I expect that it can be potentially used for building executables too.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets (hunk 4 above)
Comment (akoeplinger) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:179]: btw. these paths should be quoted since they could in theory have spaces in them

Comment (ivanpovazan) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets]: Thank you, address in the latest commit.

Comment (ivanpovazan) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets]: Thank you, address in the latest commit.

**Comment by ivanpovazan** [comment]
> Is `_iOSLibraryMode` the right name for this then? Based on your description, it is a mode without dependency on Xamarin SDK or Apple builder. We happen to use for building dynamic libraries in this PR, but I expect that it can be potentially used for building executables too.

Could you please explain how would building an iOS executable look like in such case and how we would bundle it?

**Comment by steveisok** [comment]
> > Is `_iOSLibraryMode` the right name for this then? Based on your description, it is a mode without dependency on Xamarin SDK or Apple builder. We happen to use for building dynamic libraries in this PR, but I expect that it can be potentially used for building executables too.
> 
> Could you please explain how would building an iOS executable look like in such case and how we would bundle it?

My opinion, we could do lightweight apps in theory, but given what the xam ios repo has, I can't think of any value doing it here. Maybe to align architecturally what comes out of runtime vs sdk's, but that's highly subjective.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets [hunk 5]
```diff
@@ -57,6 +57,51 @@ The .NET Foundation licenses this file to you under the MIT license.
       <StandaloneGCSupportName Condition="'$(IlcStandaloneGCSupport)' == 'true'">libstandalonegc-enabled</StandaloneGCSupportName>
     </PropertyGroup>
 
+    <PropertyGroup Condition="'$(_iOSLibraryMode)' == 'true'">
```
Comment (steveisok) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:60]: Should this be in a stand alone task?

Comment (ivanpovazan) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:60]: > Should this be in a stand alone task? At least in some form?

Do you mean to group these in a separate MSBuild target? 
- We could, but then we would have to escape setting up the following as they would occur after the new target https://github.com/dotnet/runtime/blob/b0e7ee7c8d165e49b99113e23e08d2d1c17f88a2/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets#L43-L47

- Additionally, this new target (lets call it `SetupAppleSpecificProps`) should be executed before `SetupOSSpecificProps` and I already had some attempts to make that work properly see: discussion above https://github.com/dotnet/runtime/pull/93658#discussion_r1364994941 
Basically the problem is that we already have some external targets hooking in before `SetupOSSpecificProps`, so if it happens that these external targets are also trying to set properties before `SetupAppleSpecificProps` there is no guarantee that the external target will be scheduled by MSBuild before this newly introduced one.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets [hunk 6]
```diff
@@ -30,6 +30,7 @@ The .NET Foundation licenses this file to you under the MIT license.
     <RunILLink>false</RunILLink>
     <_IsiOSLikePlatform Condition="'$(_targetOS)' == 'maccatalyst' or $(_targetOS.StartsWith('ios')) or $(_targetOS.StartsWith('tvos'))">true</_IsiOSLikePlatform>
     <_IsApplePlatform Condition="'$(_targetOS)' == 'osx' or '$(_IsiOSLikePlatform)' == 'true'">true</_IsApplePlatform>
+    <_iOSLibraryMode Condition="'$(_IsiOSLikePlatform)' == 'true' and !$(TargetFramework.Contains('$(_targetOS)')) and '$(NativeLib)' != '' and '$(CustomNativeMain)' != 'true'">true</_iOSLibraryMode>
```
Comment (am11) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:33]: Library Mode term sounds too special in this context. Can this condition be simplified to `'$(_IsiOSLikePlatform)' == 'true' && '$(NativeLib)' == 'Shared'` and inline it at call-site?  `static` and `static-pie` libs are a bit different.

Comment (steveisok) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:33]: Is there a better umbrella name to refer to when we want to consume a .net library from native?

Comment (am11) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:33]: With `<OutputType>Exe` and `<PublishAot>true` in csproj, dotnet-publish generates a native executable. With `<OutputType>Shared` and `<PublishAot>true` we get  a shared library. We can additionally set `<NativeLib>Static` to get an static lib.

For platform specialization, we have targetXX and isXX type properties in place. Unless "Library Mod"e in `nativeaot/BuildIntegration` context is bringing something very unique (which, going by the first bullet in top comment, doesn't seem to be the case), it's better to use the existing / nearby convention.

Comment (am11) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:33]: With `<OutputType>Exe` and `<PublishAot>true` in csproj, dotnet-publish generates a native executable. With `<OutputType>Library` and `<PublishAot>true` we get  a shared library. We can additionally set `<NativeLib>Static` to get a static lib.

For platform specialization, we have targetXX and isXX type properties in place. Unless "Library Mod"e in `nativeaot/BuildIntegration` context is bringing something very unique (which, going by the first bullet in top comment, doesn't seem to be the case), it's better to use the existing / nearby convention.

Comment (MichalStrehovsky) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:33]: I agree with what @jkotas and @am11 are saying here. `_iOSLibraryMode` is introducing a new concept and the question is if a new concept is avoidable.

The way we typically do checks here is that we spell out if something is platform specific or library mode specific. Like so:

https://github.com/dotnet/runtime/blob/6f4f268d0b46fa5dbcee46b7220919defb148127/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets#L384

https://github.com/dotnet/runtime/blob/6f4f268d0b46fa5dbcee46b7220919defb148127/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets#L69

(Plenty of other examples. We don't introduce a new property because the condition is still concise and readable.)

The conditions when `_iOSLibraryMode` is true are really difficult to parse for me. I wouldn't actually be able to quickly tell when this is true. For example, why do we need to condition on `!$(TargetFramework.Contains('$(_targetOS)'))` and `CustomNativeMain` and why those decide whether we pass `--target` to the linker (try to mentally inline the definition of `_iOSLibraryMode` into places that use the property and whether such definition really makes sense there).

In general - in the places that _use_ the new `_iOSLibraryMode` property, the question is why we're not keying off of `_IsiOSLikePlatform` and `NativeLib` (if necessary - I don't think it's necessary to key off the latter in many of the places that use `_iOSLibraryMode` right now - I think that goes back to Jan's hypothetical "what if we're building an exe" scenario - we don't need to have that scenario but it actually makes the conditions simpler to reason about because it makes it clear what part are due to platform differences and what parts are due to output type differences).

**Comment by jkotas** [comment]
> Yes, when building executables, apple SDK, target triple and all apple-specific build parameters, are resolved by Xamarin SDK and tooling that creates an application bundle.
> We have our internal AppleAppBuilder https://github.com/dotnet/runtime/tree/main/src/tasks/AppleAppBuilder having a minimal set of such capabilities, which we use for CI and local testing.

Right. This PR is copying some logic that exists in the Xamarin SDK or AppleAppBuilder to native AOT build targets.

> Could you please explain how would building an iOS executable look like in such case and how we would bundle it?

We would copy more logic that exists in the Xamarin SDK or AppleAppBuilder to native AOT build targets. I can imagine that this can be useful for app types (games?) that do not want to use Objective C interop provided by Xamarin SDK. I am not suggesting that this PR implements. I am just suggesting that the changes in this PR work in this direction instead of introducing a new very specialized concept.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets (hunk 6 above)
Comment (ivanpovazan) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:33]: Thank you all for you feedback and I understand the concerns.

The specific conditions (like `!$(TargetFramework.Contains('$(_targetOS)'))`) were introduced to avoid looking up the SDK and resolving relevant properties twice (when the build actually pulls in Xamarin SDK). 

I will try to simplify this.

**Comment by ivanpovazan** [comment]
> We would copy more logic that exists in the Xamarin SDK or AppleAppBuilder to native AOT build targets. I can imagine that this can be useful for app types (games?) that do not want to use Objective C interop provided by Xamarin SDK. I am not suggesting that this PR implements it. I am just suggesting that the changes in this PR work in this direction instead of introducing a new very specialized concept.

Thank you for the explanation. 

Would it make sense to invest in refactoring `AppleAppBuilder` and productize it instead (not as part of this PR)? Duplicating the logic especially inlined in MSBuild targets is error prone and would be cleaner to implement in a custom build task (which AppleAppBuilder already is).

/cc: @steveisok

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets (hunk 6 above)
Comment (MichalStrehovsky) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:33]: How expensive is running the commands? At the end of the day we're going to run an aot compiler so if it's not costly in relative terms to aot compilation I would not try to avoid it too hard.

I now understand that this is trying to approximate the exact conditions under which xamarin sdk or appbuilder would take over but that's also what makes it feel a bit like mystery code.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets [hunk 7]
```diff
@@ -57,6 +57,51 @@ The .NET Foundation licenses this file to you under the MIT license.
...
+      <AppleMinOSVersion Condition="'$(AppleMinOSVersion)' == '' and ($(_targetOS.StartsWith('ios')) or $(_targetOS.StartsWith('tvos')))">11.0</AppleMinOSVersion>
+
+      <_AppleSdkName Condition="'$(_targetOS)' == 'ios'">iphoneos</_AppleSdkName>
+      <_AppleSdkName Condition="'$(_targetOS)' == 'iossimulator'">iphonesimulator</_AppleSdkName>
+      <_AppleSdkName Condition="'$(_targetOS)' == 'tvos'">appletvos</_AppleSdkName>
+      <_AppleSdkName Condition="'$(_targetOS)' == 'tvossimulator'">appletvsimulator</_AppleSdkName>
+      <_AppleSdkName Condition="'$(_targetOS)' == 'maccatalyst'">macosx</_AppleSdkName>
+
+      <_AppleTripleArch Condition="'$(_targetArchitecture)' == 'x64'">x86_64</_AppleTripleArch>
+      <_AppleTripleArch Condition="'$(_targetArchitecture)' == 'arm64'">arm64</_AppleTripleArch>
```
Comment (am11) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:83]: Are these different than `CrossCompileArch` resolved in the previous property group? If not, lets use the existing properties.

slightly related: line 36 can be simplified as:

```diff
-       <CrossCompileRid Condition="$(_hostOS) != $(_originalTargetOS) or !$(RuntimeIdentifier.EndsWith('-$(_hostArchitecture)'))">$(RuntimeIdentifier)</CrossCompileRid>
+       <CrossCompileRid Condition="'$(_hostOS)' != '$(_originalTargetOS)' or '$(_hostArchitecture)' != '$(_targetArchitecture)'">$(RuntimeIdentifier)</CrossCompileRid>
```

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets (hunk 6 above)
Comment (ivanpovazan) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]: I have simplified the logic as suggested. Please give it another look.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets (hunk 7 above)
Comment (ivanpovazan) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets]: It was an oversight, you are right, included in my last commit. Thank you!

File: src/tests/FunctionalTests/iOS/Device/LibraryMode/iOS.Device.LibraryMode.Test.csproj [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  <!-- TODO: This should be moved to a common build integration targets to support local builds when https://github.com/dotnet/runtime/issues/93447 is addressed -->
+  <Target Name="_PrepareSetupForLocalBuild" BeforeTargets="SetupProperties" Condition="'$(UseNativeAOTRuntime)' == 'true' and $(BuildTestsOn) == 'local'">
+    <PropertyGroup>
+      <IlcToolsPath>$(CoreCLRCrossILCompilerDir)</IlcToolsPath>
+      <IlcSdkPath>$(CoreCLRAotSdkDir)</IlcSdkPath>
+      <IlcFrameworkPath>$(LibrariesAllBinArtifactsPath)</IlcFrameworkPath>
+      <IlcFrameworkNativePath>$(LibrariesAllBinArtifactsPath)</IlcFrameworkNativePath>
+    </PropertyGroup>
+    <ItemGroup>
+      <DirectPInvoke Include="__Internal" KeepDuplicates="false" />
```
Comment (am11) [inline src/tests/FunctionalTests/iOS/Device/LibraryMode/iOS.Device.LibraryMode.Test.csproj:51]: nit: since .NET 7, `System.Runtime.InteropServices.NativeLibrary.GetMainProgramHandle()` returns the self pointer in mono and coreclr alike (which we can use in native library resolver, followed by `DllImport("AnyNameOfChoiceUsedInResolver")`).

**Review by am11 — APPROVED** [review]
Two optional nits, otherwise looks great! 👍

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets [hunk 9]
```diff
@@ -57,6 +57,48 @@ The .NET Foundation licenses this file to you under the MIT license.
...
+    <PropertyGroup Condition="'$(_IsiOSLikePlatform)' == 'true'">
+      <Xcrun Condition="'$(Xcrun)' == ''">xcrun</Xcrun>
+      <_WhereXcrun>0</_WhereXcrun>
+    </PropertyGroup>
+
+    <Exec Command="command -v &quot;$(Xcrun)&quot;" IgnoreExitCode="true" StandardOutputImportance="Low" Condition="'$(_IsiOSLikePlatform)' == 'true'">
+      <Output TaskParameter="ExitCode" PropertyName="_WhereXcrun" />
+    </Exec>
+    <Error Condition="'$(_WhereXcrun)' != '0' and '$(_IsiOSLikePlatform)' == 'true'"
+      Text="'$(Xcrun)' not found in PATH. Make sure '$(Xcrun)' is available in PATH." />
```
[the remark this answers is not in the document]
Comment (ivanpovazan) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:69]: Thank you, I will look into this asap.

File: src/tests/FunctionalTests/iOS/Device/LibraryMode/iOS.Device.LibraryMode.Test.csproj (hunk 8 above)
Comment (MichalPetryka) [reply src/tests/FunctionalTests/iOS/Device/LibraryMode/iOS.Device.LibraryMode.Test.csproj:51]: > nit: since .NET 7, `System.Runtime.InteropServices.NativeLibrary.GetMainProgramHandle()` returns the self pointer in mono and coreclr alike (which we can use in native library resolver, followed by `DllImport("AnyNameOfChoiceUsedInResolver")`).

Won't that behave like a normal PInvoke, not a direct one?

Comment (am11) [reply src/tests/FunctionalTests/iOS/Device/LibraryMode/iOS.Device.LibraryMode.Test.csproj:51]: Ah, if direct P/Invoke is optimized differently (without additional frames etc.) then yes. I don't know what DirectPInvoke entails. :)

(just [remembered](https://github.com/dotnet/runtime/commit/11ca24d39b5b0aa5c31ebe037cab8915693511da) that mono returns the same cached value with both `GetMainProgramHandle()` and `_Internal`)

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets (hunk 9 above)
[the remark this answers is not in the document]
Comment (ivanpovazan) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets]: I moved the validation block for better grouping, thank you.

File: src/tests/FunctionalTests/iOS/Device/LibraryMode/iOS.Device.LibraryMode.Test.csproj (hunk 8 above)
Comment (ivanpovazan) [reply src/tests/FunctionalTests/iOS/Device/LibraryMode/iOS.Device.LibraryMode.Test.csproj]: Removed. This is not needed as the tests are not relying on direct P/Invokes. Thank you

File: src/tasks/AppleAppBuilder/Xcode.cs (hunk 1 above)
Comment (ivanpovazan) [reply src/tasks/AppleAppBuilder/Xcode.cs:349]: I have opened https://github.com/dotnet/runtime/issues/96664 to track this.

**Comment by ivanpovazan** [comment]
The CI failures are unrelated.

@steveisok @rolfbjarne @jkotas could you please take another look as we would like to get this in for `net9.0-p1`?

**Review by rolfbjarne — APPROVED** [review]
LGTM

**Review by akoeplinger — APPROVED** [review]

**Comment by jkotas** [comment]
> @jkotas could you please take another look as we would like to get this in for net9.0-p1?

The changes in the shared native AOT files LGTM.

**Review by MichalStrehovsky — COMMENTED** [review]
nativeaot/BuildIntegration looks good to me too!
